### DISCUSSION TITLE
Multiple calls to initRDKitModule

### DISCUSSION THREAD

**Discussion by fredludlow** [rdkit/rdkit-js#437] [Q&A] [answered]
Reposting from #436 

Firstly, thanks a lot for all the work that's gone into this project. It's really nice to have the rdkit functionality in the browser. I have a (hopefully) basic question, which might be answered in the docs somewhere but I couldn't find it.

If I have a webapp in which multiple independent parts might require rdkit, what's the convention around calls to initRDKitModule. Each call seems to trigger a fetch of the wasm code, so I'm guessing its setting up a new instance per-call.

I can easily define some convention within my app whereby the promise (return value of initRDKitModule) is saved on some global object or shared state, but I'm aware I might be missing something here.

**Comment by ptosco** [comment, marked as answer]
@fredludlow Hi Fred, as you already figured out, every time you call `initRDKitModule()` a new instance of the module will be created.
You may take a look at how [the `rdkit-structure-renderer` package](https://github.com/rdkit/rdkit-structure-renderer/blob/fc682615d3b5ce542ff040ca1bb046d9dc060f16/src/StructureRenderer/Renderer.js#L429) avoids multiple instantiations: `Renderer.init()` is safe to call multiple times, as it will only instantiate the RDKit module once.
You might as well find convenient to use `rdkit-structure-renderer` (it is a vanilla JS package with very few dependencies other than MinimalLib itself) rather than implementing the logic again in your app.
Note that `rdkit-structure-renderer` does not depend on `rdkit-js` - it bundles its own version of MinimalLib, which may include additional functionality compared to `rdkit-js`. `rdkit-structure-renderer` is built from the RDKit `master` branch, while `rdkit-js` follows the RDKit release cycle, hence the MinimalLib bundled in `rdkit-structure-renderer` is more bleeding edge than the one in `rdkit-js`.

**Reply by fredludlow** [reply]
Thanks @ptosco - I'm actually using it to enable client-side substructure querying rather than depiction - apart from the differences around release cycles/bleeding edge you mentioned, are there any other differences (e.g. does `rdkit-structure-renderer` exclude any functionality as not directly relevant to rendering structures?)

Somewhat relatedly, how do I configure the size of the emscripten/C++ side heap, and is there any way I can check how much is currently in use? (I'm aware I need to call `delete()` and I've not got hard numbers from profiling but I have a suspicion I could improve performance by caching the rdkit mol instances - it would be helpful to have a way to know memory usage).

**Comment by ptosco** [comment]
@fredludlow 
> are there any other differences (e.g. does rdkit-structure-renderer exclude any functionality as not directly relevant to rendering structures?)

No - `rdkit-structure-renderer` may rather include additional, optional  functionality compared to `rdkit-js`; i.e., ``rdkit-structure-renderer` will always be a superset of `rdkit-js` rather than a subset.

> Somewhat relatedly, how do I configure the size of the emscripten/C++ side heap, and is there any way I can check how much is currently in use?

To configure the size of the emscripten heap, you may use `INITIAL_MEMORY`, `MAXIMUM_MEMORY` and `ALLOW_MEMORY_GROWTH`, all of which can be configured with `emcmake` at build time. See for example this thread:
https://github.com/emscripten-core/emscripten/issues/14667
To check how much memory is available from JS, you may call `rdkitModule.HEAP8.length`. Getting memory which is currently use can be more tricky; you might have a look at this thread for some pointers:
https://github.com/emscripten-core/emscripten/issues/8082
